modules/fileout_netcdf/FONcArrayStructure.cc

```diff
@@ -0,0 +1,220 @@
+ * @throws BESInternalError if the BaseType is not a an array
+ */
+FONcArrayStructure::FONcArrayStructure(BaseType *b) :
+    FONcBaseType()
+{
+    _as = dynamic_cast<Array *>(b);
+    if (!_as) {
+        auto s = (string) "File out netcdf, array of structure was passed a " + "variable that is not an array ";
+        throw BESInternalError(s, __FILE__, __LINE__);
+    }
+    if (_as->var()->type() != dods_structure_c) {
+        string s = (string) "File out netcdf, array of structure was passed a " + "variable that is not a structure ";
+        throw BESInternalError(s, __FILE__, __LINE__);
+    }
+}
+
+/** @brief Destructor that cleans up the array of structure
+ *
+ * Delete each of the FONcBaseType instances that is a part of this
+ * structure.
+ */
+FONcArrayStructure::~FONcArrayStructure()
+{
+
+    // Adapt the code from FONcStructure.cc
+    bool done = false;
+    while (!done) {
+        // _vars are vector<FONcArrayStructureField *>
+        auto i = _vars.begin();
+        auto e = _vars.end();
+        if (i == e) {
+            done = true;
+        }
+        else {
+            FONcArrayStructureField *b = (*i);
+            delete b;
+            _vars.erase(i);
+        }
+    }
+
+}
+
+/** @brief Creates the FONc objects for each variable of the DAP array of structure
+ *
+ * For each of the variables of the DAP Structure we convert to a
+ * similar FONc object. We still follow the current way to map structure to netCDF-3.
+ * We must flatten out the structures. To do this, we embed the name of the
+ * structure as part of the name of the children variables. For example,
+ * if the structure, called s1, contains an array called a1 and an int
+ * called i1, then two variables are created in the netcdf file called
+ * s1.a1 and s1.i1.
+ *
+ * @note This method only converts the variables that are to be sent. This keeps
+ * the convert() and write() methods below from operating on DAP variables
+ * that should not be sent.
+ *
+ * @param embed The parent names of this structure.
+ * @throws BESInternalError if there is a problem converting this
+ * structure
+ */
+void FONcArrayStructure::convert(vector<string> embed, bool _dap4, bool is_dap4_group){
+    set_is_dap4(_dap4);
+    FONcBaseType::convert(embed,_dap4,is_dap4_group);
+    embed.push_back(name());
+
+    auto as_v = dynamic_cast<Structure *>(_as->var());
+    if (!as_v) {
+        string s = (string) "File out netcdf, write an array of structure was passed a " + "variable that is not a structure";
+        throw BESInternalError(s, __FILE__, __LINE__);
+    }
+
+    if (false == isNetCDF4_ENHANCED()) {
+        throw BESInternalError("Fileout netcdf: We don't support array of structure for the classical model now. ",
+                               __FILE__, __LINE__);
+    }
+
+    Constructor::Vars_iter vi = as_v->var_begin();
+    Constructor::Vars_iter ve = as_v->var_end();
+    for (; vi != ve; vi++) {
```

Comment:
I think Constructor has a method variables() that will enable th erange-base for loop, so you can write

for (auto &bt: variables()) {
    BESDEBUG("fonc", "FONcArrayStructure::convert - converting " << bt->name() << endl);
    auto fsf = new FONcArrayStructureField(bt, _as);
    vars.push_back(fsf);
    fsf->convert(embed,_dap4,is_dap4_group);
}

I think the variables() method returns a vector<BaseType*> so the type of 'bt' will be BaseType*.

Reply:
Your way is more elegant. Adopt your suggestion. Change the code. Thanks